Implement CookieStoreManager for the CookieStore API

RupinMittal opened the pull request:
#### c1a5c61946b9f75678be873346223d40f774f814
<pre>
Implement CookieStoreManager for the CookieStore API
<a href="https://rdar.apple.com/133575968">rdar://133575968</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=277892">https://bugs.webkit.org/show_bug.cgi?id=277892</a>

Reviewed by Chris Dumez.

This patch implements the CookieStoreManager which allows
Service Workers to subscribe to events on cookie changes
(<a href="https://wicg.github.io/cookie-store/#CookieStoreManager).">https://wicg.github.io/cookie-store/#CookieStoreManager).</a>

We implement the subscribe, unsubscribe, and getSubscriptions
functions. The list of subscriptions is stored in the Service
Worker Registration on the Network Process side. The WebProcess
sends IPC messages to alter or get a copy of the list.

This patch causes some WPT Layout Tests that were previously
failing to now pass. Some of the tests whose results changed
to timeout are timing out because they are testing the
functionality of the ExtendableCookieChangeEvent for Service
Workers which has not yet been implemented. Since we currently
don&apos;t plan on firing cookie change events to the Service Workers,
there&apos;s no plan to implement ExtendableCookieChangeEvent yet.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStoreManager_getSubscriptions_empty.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStoreManager_getSubscriptions_empty.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStoreManager_getSubscriptions_multiple.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStoreManager_getSubscriptions_multiple.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStoreManager_getSubscriptions_single.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStoreManager_getSubscriptions_single.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_subscribe_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_subscribe_arguments.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_subscriptions_empty.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/serviceworker_cookieStore_subscriptions_reset.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/serviceworker_cookiechange_eventhandler_mismatched_subscription.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/serviceworker_cookiechange_eventhandler_multiple_subscriptions.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/serviceworker_cookiechange_eventhandler_overlapping_subscriptions.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/serviceworker_cookiechange_eventhandler_single_subscription.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/serviceworker_oncookiechange_eventhandler_single_subscription.https.any.serviceworker-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/cookie-store/CookieChangeSubscription.h: Added.
(WebCore::CookieChangeSubscription::CookieChangeSubscription):
(WebCore::CookieChangeSubscription::isHashTableDeletedValue const):
(WTF::CookieChangeSubscriptionHash::hash):
(WTF::CookieChangeSubscriptionHash::equal):
* Source/WebCore/Modules/cookie-store/CookieStoreGetOptions.idl:
* Source/WebCore/Modules/cookie-store/CookieStoreManager.cpp:
(WebCore::CookieStoreManager::create):
(WebCore::CookieStoreManager::CookieStoreManager):
(WebCore::CookieStoreManager::subscribe):
(WebCore::CookieStoreManager::unsubscribe):
(WebCore::CookieStoreManager::getSubscriptions):
* Source/WebCore/Modules/cookie-store/CookieStoreManager.h:
* Source/WebCore/Modules/cookie-store/CookieStoreManager.idl:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/workers/service/SWClientConnection.h:
* Source/WebCore/workers/service/ServiceWorkerContainer.cpp:
(WebCore::ServiceWorkerContainer::ensureProtectedSWClientConnection):
(WebCore::ServiceWorkerContainer::addCookieChangeSubscriptions):
(WebCore::ServiceWorkerContainer::removeCookieChangeSubscriptions):
(WebCore::ServiceWorkerContainer::cookieChangeSubscriptions):
* Source/WebCore/workers/service/ServiceWorkerContainer.h:
* Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
(WebCore::ServiceWorkerRegistration::cookies):
(WebCore::ServiceWorkerRegistration::addCookieChangeSubscriptions):
(WebCore::ServiceWorkerRegistration::removeCookieChangeSubscriptions):
(WebCore::ServiceWorkerRegistration::cookieChangeSubscriptions):
(WebCore::ServiceWorkerRegistration::protectedContainer const):
* Source/WebCore/workers/service/ServiceWorkerRegistration.h:
* Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
(WebCore::WorkerSWClientConnection::~WorkerSWClientConnection):
(WebCore::WorkerSWClientConnection::matchRegistration):
(WebCore::WorkerSWClientConnection::getRegistrations):
(WebCore::WorkerSWClientConnection::whenRegistrationReady):
(WebCore::WorkerSWClientConnection::scheduleUnregisterJobInServer):
(WebCore::WorkerSWClientConnection::subscribeToPushService):
(WebCore::WorkerSWClientConnection::unsubscribeFromPushService):
(WebCore::WorkerSWClientConnection::getPushSubscription):
(WebCore::WorkerSWClientConnection::getPushPermissionState):
(WebCore::WorkerSWClientConnection::getNotifications):
(WebCore::WorkerSWClientConnection::enableNavigationPreload):
(WebCore::WorkerSWClientConnection::disableNavigationPreload):
(WebCore::WorkerSWClientConnection::setNavigationPreloadHeaderValue):
(WebCore::WorkerSWClientConnection::getNavigationPreloadState):
(WebCore::WorkerSWClientConnection::startBackgroundFetch):
(WebCore::WorkerSWClientConnection::backgroundFetchInformation):
(WebCore::WorkerSWClientConnection::backgroundFetchIdentifiers):
(WebCore::WorkerSWClientConnection::abortBackgroundFetch):
(WebCore::WorkerSWClientConnection::matchBackgroundFetch):
(WebCore::WorkerSWClientConnection::retrieveRecordResponse):
(WebCore::WorkerSWClientConnection::retrieveRecordResponseBody):
(WebCore::WorkerSWClientConnection::addCookieChangeSubscriptions):
(WebCore::WorkerSWClientConnection::removeCookieChangeSubscriptions):
(WebCore::WorkerSWClientConnection::cookieChangeSubscriptions):
* Source/WebCore/workers/service/WorkerSWClientConnection.h:
* Source/WebCore/workers/service/server/SWServerRegistration.cpp:
(WebCore::SWServerRegistration::addCookieChangeSubscriptions):
(WebCore::SWServerRegistration::removeCookieChangeSubscriptions):
(WebCore::SWServerRegistration::cookieChangeSubscriptions const):
* Source/WebCore/workers/service/server/SWServerRegistration.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::sharedPreferencesForWebProcess const):
(WebKit::WebSWServerConnection::addCookieChangeSubscriptions):
(WebKit::WebSWServerConnection::removeCookieChangeSubscriptions):
(WebKit::WebSWServerConnection::cookieChangeSubscriptions):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::WebSWClientConnection::addCookieChangeSubscriptions):
(WebKit::WebSWClientConnection::removeCookieChangeSubscriptions):
(WebKit::WebSWClientConnection::cookieChangeSubscriptions):
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.h:

Canonical link: <a href="https://commits.webkit.org/282417@main">https://commits.webkit.org/282417@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4592fa1a835293d934af318a2503689627e0c665

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/63140 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/42496 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/15736 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/67161 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/13748 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/65260 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/50183 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/14028 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/50878 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/9490 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/66209 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/39463 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/54670 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/31564 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/62651 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/36148 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/12017 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/12620 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/56250 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/57688 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/12346 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/68856 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/62383 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/7086 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/11958 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/58190 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/7118 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/54742 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/58403 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13989 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/5909 "Passed tests") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/84146 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/38316 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/14827 "Found 1473 new JSC stress test failures: microbenchmarks/memcpy-wasm-large.js.no-cjit-validate-phases, microbenchmarks/memcpy-wasm-medium.js.no-cjit-validate-phases, microbenchmarks/memcpy-wasm-small.js.no-cjit-validate-phases, microbenchmarks/memcpy-wasm.js.no-cjit-validate-phases, microbenchmarks/wasm-cc-int-to-int.js.default-wasm, stress/proxy-get-and-set-recursion-stack-overflow.js.eager-jettison-no-cjit, stress/proxy-get-and-set-recursion-stack-overflow.js.no-cjit-collect-continuously, stress/proxy-get-and-set-recursion-stack-overflow.js.no-cjit-validate-phases, stress/proxy-set.js.bytecode-cache, stress/proxy-set.js.default ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/39396 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/40507 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/39138 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
```diff
@@ -762,6 +763,54 @@ void WebSWServerConnection::retrieveRecordResponseBody(WebCore::BackgroundFetchR
+    callback({ });
+}
+
+void WebSWServerConnection::removeCookieChangeSubscriptions(WebCore::ServiceWorkerRegistrationIdentifier registrationIdentifier, Vector<WebCore::CookieChangeSubscription>&& subscriptions, ExceptionOrVoidCallback&& callback)
```
This is the Network Process side function where the test is failing.

RupinMittal replied:
We spoke offline about this and determined that the spec is not clear about this case. Since the spec is not clear and the WPT expect no error here, we added a FIXME and throw no error. We did file an issue on the spec for clarification and included the link in the FIXME as well (https://github.com/WICG/cookie-store/issues/237)

szewai commented on Source/WebCore/workers/service/server/SWServerRegistration.cpp:
```diff
@@ -33,6 +33,7 @@
 #include "SWServerWorker.h"
 #include "ServiceWorkerTypes.h"
 #include "ServiceWorkerUpdateViaCache.h"
+#include <wtf/Vector.h>
```
Why do we include this header in both .h and .cpp?

RupinMittal replied:
Fixed to include `Forward.h` in the header instead.

szewai commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.cpp:
```diff
@@ -27,34 +27,101 @@
 #include "CookieStoreManager.h"
 
 #include "CookieStoreGetOptions.h"
+#include "JSCookieStoreGetOptions.h"
```
What is this for?

RupinMittal replied:
Not needed here--fixed!

szewai commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.idl:
```diff
@@ -30,7 +30,7 @@
     SecureContext
 ] interface CookieStoreManager {
-    Promise<undefined> subscribe(sequence<CookieStoreGetOptions> subscriptions);
+    [CallWith=CurrentScriptExecutionContext] Promise<undefined> subscribe(sequence<CookieStoreGetOptions> subscriptions);
```
The `CallWith` is unnecessary; `subscribe()` can use the `ScriptExecutionContext` of `ServiceWorkerRegistration`.

RupinMittal replied:
Fixed!

szewai commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.idl:
```diff
@@ -30,7 +30,7 @@
 ] interface CookieStoreManager {
-    Promise<undefined> subscribe(sequence<CookieStoreGetOptions> subscriptions);
+    [CallWith=CurrentScriptExecutionContext] Promise<undefined> subscribe(sequence<CookieStoreGetOptions> subscriptions);
+    [CallWith=CurrentScriptExecutionContext] Promise<undefined> unsubscribe(sequence<CookieStoreGetOptions> subscriptions);
```
Ditto.

RupinMittal replied:
Fixed!

szewai commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.cpp:
```diff
@@ -27,34 +27,101 @@
 {
-    promise->reject(ExceptionCode::NotSupportedError);
+    if (!m_serviceWorkerRegistration) {
+        promise->reject(ExceptionCode::SecurityError);
```
Let's return `InvalidStateError` and with error message. `SecurityError` usually means the operation is not safe.

RupinMittal replied:
Fixed!

szewai commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.cpp:
```diff
@@ -27,34 +27,101 @@
+
+        if (url.substring(0, scope.length()) != scope) {
+            promise->reject(ExceptionCode::TypeError);
+            return;
```
`String` has `startsWith()` function

RupinMittal replied:
Fixed!

szewai commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.cpp:
```diff
@@ -27,34 +27,101 @@
 {
-    promise->reject(ExceptionCode::NotSupportedError);
+    if (!m_serviceWorkerRegistration) {
+        promise->reject(ExceptionCode::SecurityError);
```
`InvalidStateError`

RupinMittal replied:
Fixed!

szewai commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.cpp:
```diff
@@ -27,34 +27,101 @@
+
+    Vector<CookieChangeSubscription> cookieChangeSubscriptions;
+    for (auto& subscription : subscriptions) {
+        auto url = URL(context.url(), subscription.url).string();
```
The null check (`subscription.url.isNull()`) should be performed first, so that if it's null we don't need to do the parse.

Also you should use `ServiceWorkerRegistration`'s `scriptExecutionContext()`, and move this validation to `ServiceWorkerRegistration::addCookieChangeSubscriptions()`.

Same for the other functions.

RupinMittal replied:
Fixed!

szewai commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.cpp:
```diff
@@ -27,34 +27,101 @@
 {
-    promise->reject(ExceptionCode::NotSupportedError);
+    if (!m_serviceWorkerRegistration) {
+        promise->reject(ExceptionCode::SecurityError);
```
`InvalidStateError`

RupinMittal replied:
Fixed!

szewai commented on Source/WebCore/workers/service/ServiceWorkerContainer.cpp:
```diff
@@ -36,6 +38,7 @@
 #include "Exception.h"
 #include "FrameLoader.h"
 #include "IDLTypes.h"
+#include "JSCookieStoreGetOptions.h"
```
What is this for?

RupinMittal replied:
This is needed to resolve the promise in `ServiceWorkerContainer::cookieChangeSubscriptions`

szewai commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -352,8 +352,23 @@ void ServiceWorkerRegistration::getNotifications(const GetNotificationOptions& f
 
+void ServiceWorkerRegistration::addCookieChangeSubscriptions(Vector<CookieChangeSubscription>&& subscriptions, Ref<DeferredPromise>&& promise)
+{
+    m_container->addCookieChangeSubscriptions(identifier(), WTFMove(subscriptions), WTFMove(promise));
```
Let's reject promise `if (isContextStopped())` (script execution context stops running)

RupinMittal replied:
Fixed! Since `isContextStopped()` also checks if the `ScriptExecutionContext` is not null, I removed the check for that and use only `isContextStopped()`

szewai commented on Source/WebCore/workers/service/ServiceWorkerContainer.cpp:
```diff
@@ -761,4 +765,39 @@ void ServiceWorkerContainer::getNavigationPreloadState(ServiceWorkerRegistration
+{
+    ensureSWClientConnection().addCookieChangeSubscriptions(identifier, WTFMove(subscriptions), [promise = WTFMove(promise)](auto&& result) mutable {
+        if (result.hasException())
+            promise->reject(result.exception().code());
```
We should to create exception with both error code and message.

RupinMittal replied:
Fixed!

szewai reviewed (commented):
I took a pass on the client-side implementation and found some issues. 
Let's fix the build error and test errors first.

cdumez commented on Source/WebCore/workers/service/SWClientConnection.h:
```diff
@@ -141,6 +143,12 @@ class SWClientConnection : public RefCounted<SWClientConnection> {
     using RetrieveRecordResponseBodyCallback = Function<void(Expected<RefPtr<SharedBuffer>, ResourceError>&&)>;
     virtual void retrieveRecordResponseBody(BackgroundFetchRecordIdentifier, RetrieveRecordResponseBodyCallback&&) = 0;
 
+
```
nit: one too many blank lines here?

cdumez commented on Source/WebCore/workers/service/ServiceWorkerContainer.cpp:
```diff
@@ -761,4 +765,39 @@ void ServiceWorkerContainer::getNavigationPreloadState(ServiceWorkerRegistration
+        else {
+            auto cookieChangeSubscriptionList = result.releaseReturnValue();
+            Vector<CookieStoreGetOptions> subscriptions;
+            for (auto& subscription : cookieChangeSubscriptionList)
```
This should likely use `WTF::map()`

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -352,8 +353,67 @@ void ServiceWorkerRegistration::getNotifications(const GetNotificationOptions& f
+        if (subscription.url.isNull())
+            url = scope();
+        else
+            url = URL(scriptExecutionContext()->url(), subscription.url).string();
```
You likely want to use ScriptExecutionContext::completeURL()

RupinMittal replied:
I tried using `completeURL()` originally but encountered an issue. The WPT test `/LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_subscribe_arguments.https.any.js` has a test that calls `CookieStore::Subscribe` without a url and expects it to pass. Since this test is passing in a null url, passing it to `completeURL()` resulted in a null result for the url and as per the spec, I returned a TypeError. Since the spec and test seemed to be conflicting, I uploaded this patch as a draft and asked about this. Sihui responded here (https://github.com/WebKit/WebKit/pull/31979#issuecomment-2279485767). She also mentioned that we should file a spec issue about this and use a FIXME for now. Which is what I did and is why we need to use `URL()` here. Here is the spec issue we filed: https://github.com/WICG/cookie-store/issues/236

cdumez replied:
You're dealing with the null URL above:
```
if (subscription.url.isNull())
            url = scope();
```
So I don't understand your answer.

RupinMittal replied:
Oh right--thank you for pointing that out. Fixed!

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -352,8 +353,67 @@ void ServiceWorkerRegistration::getNotifications(const GetNotificationOptions& f
+
+        if (!url.startsWith(scope())) {
+            promise->reject(ExceptionCode::TypeError);
+                return;
```
indentation issue.

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -352,8 +353,67 @@ void ServiceWorkerRegistration::getNotifications(const GetNotificationOptions& f
 
+void ServiceWorkerRegistration::addCookieChangeSubscriptions(Vector<CookieStoreGetOptions>&& subscriptions, Ref<DeferredPromise>&& promise)
+{
+    Vector<CookieChangeSubscription> cookieChangeSubscriptions;
```
We can probably:
```
cookieChangeSubscriptions.reserveInitialCapacity(subscriptions.size());
```

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -352,8 +353,67 @@ void ServiceWorkerRegistration::getNotifications(const GetNotificationOptions& f
+
+void ServiceWorkerRegistration::removeCookieChangeSubscriptions(Vector<CookieStoreGetOptions>&& subscriptions, Ref<DeferredPromise>&& promise)
+{
+    Vector<CookieChangeSubscription> cookieChangeSubscriptions;
```
ditto on reserving capacity.

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -352,8 +353,67 @@ void ServiceWorkerRegistration::getNotifications(const GetNotificationOptions& f
+        if (subscription.url.isNull())
+            url = scope();
+        else
+            url = URL(scriptExecutionContext()->url(), subscription.url).string();
```
should use completeURL()

RupinMittal replied:
Made a comment about this above.

cdumez replied:
I still think this should use completeURL().

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -352,8 +353,67 @@ void ServiceWorkerRegistration::getNotifications(const GetNotificationOptions& f
+
+        if (!url.startsWith(scope())) {
+            promise->reject(ExceptionCode::TypeError);
+                return;
```
indentation issue.

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -352,8 +353,67 @@ void ServiceWorkerRegistration::getNotifications(const GetNotificationOptions& f
+
+void ServiceWorkerRegistration::cookieChangeSubscriptions(Ref<DeferredPromise>&& promise)
+{
+    m_container->cookieChangeSubscriptions(identifier(), WTFMove(promise));
```
Please follow smart pointer guidelines. You need a Ref/RefPtr to the container *on the stack*.

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -352,8 +353,67 @@ void ServiceWorkerRegistration::getNotifications(const GetNotificationOptions& f
+        cookieChangeSubscriptions.append({ subscription.name, url });
+    }
+
+    m_container->removeCookieChangeSubscriptions(identifier(), WTFMove(cookieChangeSubscriptions), WTFMove(promise));
```
Please follow smart pointer guidelines. You need a Ref/RefPtr to the container *on the stack*.

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -352,8 +353,67 @@ void ServiceWorkerRegistration::getNotifications(const GetNotificationOptions& f
+        cookieChangeSubscriptions.append({ subscription.name, url });
+    }
+
+    m_container->addCookieChangeSubscriptions(identifier(), WTFMove(cookieChangeSubscriptions), WTFMove(promise));
```
Please follow smart pointer guidelines. You need a Ref/RefPtr to the container *on the stack*.

cdumez commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -114,6 +116,10 @@ WorkerSWClientConnection::~WorkerSWClientConnection()
     for (auto& callback : retrieveRecordResponseBodyCallbacks.values())
         callback(makeUnexpected(ResourceError { errorDomainWebKitInternal, 0, { }, "context stopped"_s }));
+
+    auto cookieChangeSubscriptionsCallbacks = WTFMove(m_cookieChangeSubscriptionsCallback);
```
`std::exchange()` would be a bit cleaner here.

cdumez commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -551,4 +557,52 @@ void WorkerSWClientConnection::retrieveRecordResponseBody(BackgroundFetchRecordI
 
+void WorkerSWClientConnection::addCookieChangeSubscriptions(ServiceWorkerRegistrationIdentifier registrationIdentifier, Vector<CookieChangeSubscription>&& subscriptions, ExceptionOrVoidCallback&& callback)
+{
+    uint64_t requestIdentifier = ++m_lastRequestIdentifier;
```
Per safe C++ guidelines, we should stop using weakly typed identifiers.

cdumez commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -551,4 +557,52 @@ void WorkerSWClientConnection::retrieveRecordResponseBody(BackgroundFetchRecordI
+    m_voidCallbacks.add(requestIdentifier, WTFMove(callback));
+
+    callOnMainThread([thread = m_thread, requestIdentifier, registrationIdentifier, subscriptions = crossThreadCopy(WTFMove(subscriptions))]() mutable {
+        auto& connection = ServiceWorkerProvider::singleton().serviceWorkerConnection();
```
connection needs a smart pointer.

cdumez commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -551,4 +557,52 @@ void WorkerSWClientConnection::retrieveRecordResponseBody(BackgroundFetchRecordI
+        connection.addCookieChangeSubscriptions(registrationIdentifier, WTFMove(subscriptions), [thread = WTFMove(thread), requestIdentifier](auto&& result) {
+            thread->runLoop().postTaskForMode([requestIdentifier, result = crossThreadCopy(WTFMove(result))](auto& scope) mutable {
+                auto callback = downcast<WorkerGlobalScope>(scope).swClientConnection().m_voidCallbacks.take(requestIdentifier);
+                callback(WTFMove(result));
```
Are we sure there is always a callback with the given identifier here? If not, this will crash.

cdumez commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -551,4 +557,52 @@ void WorkerSWClientConnection::retrieveRecordResponseBody(BackgroundFetchRecordI
+
+void WorkerSWClientConnection::removeCookieChangeSubscriptions(ServiceWorkerRegistrationIdentifier registrationIdentifier, Vector<CookieChangeSubscription>&& subscriptions, ExceptionOrVoidCallback&& callback)
+{
+    uint64_t requestIdentifier = ++m_lastRequestIdentifier;
```
Should use a strongly typed identifier.

cdumez commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -551,4 +557,52 @@ void WorkerSWClientConnection::retrieveRecordResponseBody(BackgroundFetchRecordI
+    m_voidCallbacks.add(requestIdentifier, WTFMove(callback));
+
+    callOnMainThread([thread = m_thread, requestIdentifier, registrationIdentifier, subscriptions = crossThreadCopy(WTFMove(subscriptions))]() mutable {
+        auto& connection = ServiceWorkerProvider::singleton().serviceWorkerConnection();
```
smart pointer.

cdumez commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -551,4 +557,52 @@ void WorkerSWClientConnection::retrieveRecordResponseBody(BackgroundFetchRecordI
+        connection.removeCookieChangeSubscriptions(registrationIdentifier, WTFMove(subscriptions), [thread = WTFMove(thread), requestIdentifier](auto&& result) {
+            thread->runLoop().postTaskForMode([requestIdentifier, result = crossThreadCopy(WTFMove(result))](auto& scope) mutable {
+                auto callback = downcast<WorkerGlobalScope>(scope).swClientConnection().m_voidCallbacks.take(requestIdentifier);
+                callback(WTFMove(result));
```
Can `callback` be null?

cdumez commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -551,4 +557,52 @@ void WorkerSWClientConnection::retrieveRecordResponseBody(BackgroundFetchRecordI
+
+void WorkerSWClientConnection::cookieChangeSubscriptions(ServiceWorkerRegistrationIdentifier registrationIdentifier, ExceptionOrCookieChangeSubscriptionsCallback&& callback)
+{
+    uint64_t requestIdentifier = ++m_lastRequestIdentifier;
```
strongly typed identifier.

cdumez commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -551,4 +557,52 @@ void WorkerSWClientConnection::retrieveRecordResponseBody(BackgroundFetchRecordI
+    m_cookieChangeSubscriptionsCallback.add(requestIdentifier, WTFMove(callback));
+
+    callOnMainThread([thread = m_thread, requestIdentifier, registrationIdentifier]() mutable {
+        auto& connection = ServiceWorkerProvider::singleton().serviceWorkerConnection();
```
`Ref`

cdumez commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -551,4 +557,52 @@ void WorkerSWClientConnection::retrieveRecordResponseBody(BackgroundFetchRecordI
+        connection.cookieChangeSubscriptions(registrationIdentifier, [thread = WTFMove(thread), requestIdentifier](auto&& result) {
+            thread->runLoop().postTaskForMode([requestIdentifier, result = crossThreadCopy(WTFMove(result))](auto& scope) mutable {
+                auto callback = downcast<WorkerGlobalScope>(scope).swClientConnection().m_cookieChangeSubscriptionsCallback.take(requestIdentifier);
+                callback(WTFMove(result));
```
can `callback` be null?

RupinMittal replied:
Added a check

cdumez commented on Source/WebCore/workers/service/server/SWServerRegistration.cpp:
```diff
@@ -423,4 +424,39 @@ std::optional<ExceptionData> SWServerRegistration::setNavigationPreloadHeaderVal
+{
+    for (auto& requestedSubscription : subscriptions) {
+        bool isNewSubscription = true;
+        for (auto& currentSubscription : m_cookieChangeSubscriptions) {
```
Feels like this loop could be replaced with `WTF::anyOf()`

RupinMittal replied:
Changed to use HashSet

cdumez commented on Source/WebCore/workers/service/server/SWServerRegistration.cpp:
```diff
@@ -423,4 +424,39 @@ std::optional<ExceptionData> SWServerRegistration::setNavigationPreloadHeaderVal
+        }
+
+        if (isNewSubscription)
+            m_cookieChangeSubscriptions.append(WTFMove(requestedSubscription));
```
Did we make sure that Vector was the right data structure for this? Seems we need to guarantee uniqueness, which is usually a case for using HashMap/HashSet/ListHashSet.

RupinMittal replied:
Good point--changed to use HashSet.

cdumez commented on Source/WebCore/workers/service/server/SWServerRegistration.cpp:
```diff
@@ -423,4 +424,39 @@ std::optional<ExceptionData> SWServerRegistration::setNavigationPreloadHeaderVal
+{
+    for (auto& requestedSubscription : subscriptions) {
+        size_t i = 0;
+        for (i = 0; i < m_cookieChangeSubscriptions.size(); i++) {
```
Seems like this could use `Vector::removeFirstMatching()` but also, not sure Vector is the right data structure for the job.

RupinMittal replied:
Changed to use HashSet.

cdumez commented on Source/WebCore/workers/service/server/SWServerRegistration.cpp:
```diff
@@ -423,4 +424,39 @@ std::optional<ExceptionData> SWServerRegistration::setNavigationPreloadHeaderVal
+    }
+}
+
+Vector<CookieChangeSubscription> SWServerRegistration::cookieChangeSubscriptions()
```
Should likely:
1. Be inline in the header
2. be const
3. return ` const Vector&`

RupinMittal replied:
I changed the subscriptions to be stored in a HashSet instead of Vector. Since this function now needs to create a Vector of the subscriptions and that Vector will be destroyed at the end of the function, I have changed the function to be `const`, but did not `inline` the function or have it return `const Vector&`. I suppose we could still `inline`, but the function is no longer just a return statement, so I don't know if that would make much sense. Let me know what you think.

cdumez commented on Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
```diff
@@ -765,6 +767,47 @@ void WebSWServerConnection::retrieveRecordResponseBody(WebCore::BackgroundFetchR
+{
+    RefPtr registration = protectedServer()->getRegistration(registrationIdentifier);
+    if (!registration) {
+        RELEASE_LOG_ERROR(CookieStore, "No registration");
```
The logging should likely mention the class/function name. `RELEASE_LOG_ERROR` doesn't do that for you, which is why we usually wrap `RELEASE_LOG*` calls in a new macro.

cdumez commented on Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
```diff
@@ -765,6 +767,47 @@ void WebSWServerConnection::retrieveRecordResponseBody(WebCore::BackgroundFetchR
+{
+    RefPtr registration = protectedServer()->getRegistration(registrationIdentifier);
+    if (!registration) {
+        RELEASE_LOG_ERROR(CookieStore, "No registration");
```
ditto about logging.

cdumez commented on Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
```diff
@@ -765,6 +767,47 @@ void WebSWServerConnection::retrieveRecordResponseBody(WebCore::BackgroundFetchR
+{
+    RefPtr registration = protectedServer()->getRegistration(registrationIdentifier);
+    if (!registration) {
+        RELEASE_LOG_ERROR(CookieStore, "No registration");
```
ditto about logging.

cdumez commented on Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.messages.in:
```diff
@@ -64,6 +64,10 @@ messages -> WebSWServerConnection NotRefCounted {
     RetrieveRecordResponse(WebCore::BackgroundFetchRecordIdentifier recordIdentifier) -> (Expected<WebCore::ResourceResponse, WebCore::ExceptionData> result)
     RetrieveRecordResponseBody(WebCore::BackgroundFetchRecordIdentifier recordIdentifier, WebKit::RetrieveRecordResponseBodyCallbackIdentifier callbackIdentifier)
 
+    AddCookieChangeSubscriptions(WebCore::ServiceWorkerRegistrationIdentifier identifier, Vector<WebCore::CookieChangeSubscription> subscriptions) -> (std::optional<WebCore::ExceptionData> result)
```
All these new IPC end points likely need to be guarded behind a `[EnabledBy=X]` for security.

RupinMittal replied:
Should I use both `CookieStoreAPIEnabled` and `CookieStoreManagerEnabled` flags or just `CookieStoreManagerEnabled` ?

RupinMittal replied:
I used just `CookieStoreManagerEnabled` since these functions are only being called from `CookieStoreManager`

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.cpp:
```diff
@@ -28,33 +28,55 @@
+        return;
+    }
+
+    m_serviceWorkerRegistration->addCookieChangeSubscriptions(WTFMove(subscriptions), WTFMove(promise));
```
We need a RefPtr to m_serviceWorkerRegistration on the stack, per smart pointer guidelines.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.cpp:
```diff
@@ -28,33 +28,55 @@
+        return;
+    }
+
+    m_serviceWorkerRegistration->removeCookieChangeSubscriptions(WTFMove(subscriptions), WTFMove(promise));
```
Need a smart pointer on the stack.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.cpp:
```diff
@@ -28,33 +28,55 @@
+        return;
+    }
+
+    m_serviceWorkerRegistration->cookieChangeSubscriptions(WTFMove(promise));
```
Need a smart pointer on the stack.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.h:
```diff
@@ -26,24 +26,68 @@
+        : name(deletedValue)
+    { }
+
+    bool operator==(const CookieChangeSubscription& other) const
```
`= default;`
No need to write your own implementation as it does the default thing AFAICT.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.h:
```diff
@@ -26,24 +26,68 @@
+
+    static bool equal(const CookieChangeSubscription& a, const CookieChangeSubscription& b)
+    {
+        return (a.name == b.name) && (a.url == b.url);
```
Why can't we use `a == b` here, you wrote an operator== above.

cdumez commented on Source/WebCore/workers/service/ServiceWorkerContainer.cpp:
```diff
@@ -761,4 +765,37 @@ void ServiceWorkerContainer::getNavigationPreloadState(ServiceWorkerRegistration
 
+void ServiceWorkerContainer::addCookieChangeSubscriptions(ServiceWorkerRegistrationIdentifier identifier, Vector<CookieChangeSubscription>&& subscriptions, Ref<DeferredPromise>&& promise)
+{
+    ensureSWClientConnection().addCookieChangeSubscriptions(identifier, WTFMove(subscriptions), [promise = WTFMove(promise)](auto&& result) mutable {
```
We need a smart pointer to the SWClientConnection on the stack before calling addCookieChangeSubscriptions() on it. I suggest introducing a new `ensureProtectedSWClientConnection()` which returns a `Ref< SWClientConnection >` and call it here.

cdumez commented on Source/WebCore/workers/service/ServiceWorkerContainer.cpp:
```diff
@@ -761,4 +765,37 @@ void ServiceWorkerContainer::getNavigationPreloadState(ServiceWorkerRegistration
+
+void ServiceWorkerContainer::removeCookieChangeSubscriptions(ServiceWorkerRegistrationIdentifier identifier, Vector<CookieChangeSubscription>&& subscriptions, Ref<DeferredPromise>&& promise)
+{
+    ensureSWClientConnection().removeCookieChangeSubscriptions(identifier, WTFMove(subscriptions), [promise = WTFMove(promise)](auto&& result) mutable {
```
ensureProtectedSWClientConnection

cdumez commented on Source/WebCore/workers/service/ServiceWorkerContainer.cpp:
```diff
@@ -761,4 +765,37 @@ void ServiceWorkerContainer::getNavigationPreloadState(ServiceWorkerRegistration
+
+void ServiceWorkerContainer::cookieChangeSubscriptions(ServiceWorkerRegistrationIdentifier identifier, Ref<DeferredPromise>&& promise)
+{
+    ensureSWClientConnection().cookieChangeSubscriptions(identifier, [promise = WTFMove(promise)](auto&& result) mutable {
```
ensureProtectedSWClientConnection

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.h:
```diff
@@ -116,6 +122,8 @@ class ServiceWorkerRegistration final : public RefCounted<ServiceWorkerRegistrat
     void refEventTarget() final { ref(); }
     void derefEventTarget() final { deref(); }
 
+    Ref<ServiceWorkerContainer> protectedContainer() const;
```
This should be `public`, right next to `container()`. Why make one public and not the other?

cdumez commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -551,4 +557,64 @@ void WorkerSWClientConnection::retrieveRecordResponseBody(BackgroundFetchRecordI
+            thread->runLoop().postTaskForMode([requestIdentifier, result = crossThreadCopy(WTFMove(result))](auto& scope) mutable {
+                auto callback = downcast<WorkerGlobalScope>(scope).swClientConnection().m_voidCallbacks.take(requestIdentifier);
+                if (!callback) {
+                    callback(Exception { ExceptionCode::SecurityError, "Missing callback"_s });
```
How is this a SecurityError? I recommend `InvalidStateError`. Also, callback is internal WebKit state so "Missing callback" is not going to be useful to the Web developer getting this error.

cdumez commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -551,4 +557,64 @@ void WorkerSWClientConnection::retrieveRecordResponseBody(BackgroundFetchRecordI
+            thread->runLoop().postTaskForMode([requestIdentifier, result = crossThreadCopy(WTFMove(result))](auto& scope) mutable {
+                auto callback = downcast<WorkerGlobalScope>(scope).swClientConnection().m_voidCallbacks.take(requestIdentifier);
+                if (!callback) {
+                    callback(Exception { ExceptionCode::SecurityError, "Missing callback"_s });
```
Same comment as above.

cdumez commented on Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
```diff
@@ -551,4 +557,64 @@ void WorkerSWClientConnection::retrieveRecordResponseBody(BackgroundFetchRecordI
+            thread->runLoop().postTaskForMode([requestIdentifier, result = crossThreadCopy(WTFMove(result))](auto& scope) mutable {
+                auto callback = downcast<WorkerGlobalScope>(scope).swClientConnection().m_cookieChangeSubscriptionsCallback.take(requestIdentifier);
+                if (!callback) {
+                    callback(Exception { ExceptionCode::SecurityError, "Missing callback"_s });
```
Same comment as above.

cdumez commented on Source/WebCore/workers/service/WorkerSWClientConnection.h:
```diff
@@ -26,12 +26,19 @@
+struct CookieChangeSubscription;
+
+struct SWClientRequestIdentifierType;
+using SWClientRequestIdentifier = AtomicObjectIdentifier<SWClientRequestIdentifierType>;
```
Probably this should be private, inside the class scope. I don't see any reason to expose this globally or even outside the class?

cdumez commented on Source/WebCore/workers/service/server/SWServerRegistration.cpp:
```diff
@@ -423,4 +424,24 @@ std::optional<ExceptionData> SWServerRegistration::setNavigationPreloadHeaderVal
 
+void SWServerRegistration::addCookieChangeSubscriptions(Vector<CookieChangeSubscription>&& subscriptions)
+{
+    for (auto& subscription : subscriptions)
```
You don't need this loop:
`m_cookieChangeSubscriptions.add(subscriptions.begin(), subscriptions.end());`

cdumez commented on Source/WebCore/workers/service/server/SWServerRegistration.cpp:
```diff
@@ -423,4 +424,24 @@ std::optional<ExceptionData> SWServerRegistration::setNavigationPreloadHeaderVal
+
+void SWServerRegistration::removeCookieChangeSubscriptions(Vector<CookieChangeSubscription>&& subscriptions)
+{
+    for (auto& subscription : subscriptions)
```
Same as above.

cdumez commented on Source/WebCore/workers/service/server/SWServerRegistration.cpp:
```diff
@@ -423,4 +424,24 @@ std::optional<ExceptionData> SWServerRegistration::setNavigationPreloadHeaderVal
+
+Vector<CookieChangeSubscription> SWServerRegistration::cookieChangeSubscriptions() const
+{
+    Vector<CookieChangeSubscription> subscriptions;
```
`return copyToVector(m_cookieChangeSubscriptions);` ?

cdumez commented on Source/WebCore/workers/service/server/SWServerRegistration.h:
```diff
@@ -133,6 +141,8 @@ class SWServerRegistration : public RefCounted<SWServerRegistration>, public Can
     RefPtr<SWServerWorker> m_waitingWorker;
     RefPtr<SWServerWorker> m_activeWorker;
 
+    HashSet<CookieChangeSubscription, CookieChangeSubscriptionHash, SimpleClassHashTraits<CookieChangeSubscription>> m_cookieChangeSubscriptions;
```
If done properly, you should not need to specify those 2 last template parameters. You're likely missing the `HashTraits` template specialization for your type. You can use NetworkCacheKey.h as example.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.h:
```diff
@@ -26,24 +26,68 @@
+class WeakPtrImplWithEventTargetData;
+struct CookieStoreGetOptions;
+
+struct CookieChangeSubscription {
```
Should likely go to its own header.

cdumez commented on Source/WebKit/Scripts/webkit/messages.py:
```diff
@@ -848,6 +848,7 @@ def headers_for_type(type):
         'WebCore::CompositeMode': ['<WebCore/GraphicsTypes.h>'],
         'WebCore::CompositeOperator': ['<WebCore/GraphicsTypes.h>'],
         'WebCore::Cookie': ['<WebCore/Cookie.h>'],
+        'WebCore::CookieChangeSubscription': ['<WebCore/CookieStoreManager.h>'],
```
May not be needed anymore once you move CookieChangeSubscription to its own CookieChangeSubscription.h header.

cdumez commented on Source/WebKit/Scripts/webkit/messages.py:
```diff
@@ -848,6 +848,7 @@ def headers_for_type(type):
         'WebCore::CompositeMode': ['<WebCore/GraphicsTypes.h>'],
         'WebCore::CompositeOperator': ['<WebCore/GraphicsTypes.h>'],
         'WebCore::Cookie': ['<WebCore/Cookie.h>'],
```
I doubt this one is needed since the type and header name match 🤷🏻

cdumez commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -6676,6 +6676,12 @@ struct WebCore::CookieStoreGetOptions {
 };
 
+header: <WebCore/CookieStoreManager.h>
+[CustomHeader] struct WebCore::CookieChangeSubscription {
```
Please drop CustomHeader once you move to own header.

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.cpp:
```diff
@@ -28,33 +28,53 @@
+void CookieStoreManager::subscribe(Vector<CookieStoreGetOptions>&& subscriptions, Ref<DeferredPromise>&& promise)
 {
-    promise->reject(ExceptionCode::NotSupportedError);
+    if (RefPtr<ServiceWorkerRegistration> registration = m_serviceWorkerRegistration.get()) {
```
`<ServiceWorkerRegistration>` is unnecessary

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.cpp:
```diff
@@ -28,33 +28,53 @@
+void CookieStoreManager::unsubscribe(Vector<CookieStoreGetOptions>&& subscriptions, Ref<DeferredPromise>&& promise)
 {
-    promise->reject(ExceptionCode::NotSupportedError);
+    if (RefPtr<ServiceWorkerRegistration> registration = m_serviceWorkerRegistration.get()) {
```
`<ServiceWorkerRegistration>` is unnecessary

cdumez commented on Source/WebCore/Modules/cookie-store/CookieStoreManager.cpp:
```diff
@@ -28,33 +28,53 @@
+void CookieStoreManager::getSubscriptions(Ref<DeferredPromise>&& promise)
 {
-    promise->reject(ExceptionCode::NotSupportedError);
+    if (RefPtr<ServiceWorkerRegistration> registration = m_serviceWorkerRegistration.get()) {
```
`<ServiceWorkerRegistration>` is unnecessary

cdumez commented on Source/WebCore/workers/service/ServiceWorkerContainer.cpp:
```diff
@@ -600,6 +605,19 @@ SWClientConnection& ServiceWorkerContainer::ensureSWClientConnection()
+{
+    ASSERT(scriptExecutionContext());
+    if (!m_swConnection || m_swConnection->isClosed()) {
+        // Using RefPtr here results in an m_adoptionIsRequired assert.
```
We should call `relaxAdoptionRequirement();` in the constructor of the object that's triggering this. We want to be using assert pointers.

also, this function should just call `ensureSWClientConnection()`, not duplicate it..

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -352,8 +355,74 @@ void ServiceWorkerRegistration::getNotifications(const GetNotificationOptions& f
+            url = scriptExecutionContext()->completeURL(subscription.url).string();
+
+        if (!url.startsWith(scope())) {
+            promise->reject(ExceptionCode::TypeError);
```
Please provide a useful error message to the developer.

cdumez replied:
Also, it seems we only need to do this check in the `else` case above.

cdumez commented on Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
```diff
@@ -352,8 +355,74 @@ void ServiceWorkerRegistration::getNotifications(const GetNotificationOptions& f
+            url = scriptExecutionContext()->completeURL(subscription.url).string();
+
+        if (!url.startsWith(scope())) {
+            promise->reject(ExceptionCode::TypeError);
```
Same as above.

cdumez reviewed (commented):
Almost there.

cdumez reviewed (approved):
LGTM

RupinMittal committed:
Implement CookieStoreManager for the CookieStore API
rdar://133575968
https://bugs.webkit.org/show_bug.cgi?id=277892

Reviewed by Chris Dumez.

This patch implements the CookieStoreManager which allows
Service Workers to subscribe to events on cookie changes
(https://wicg.github.io/cookie-store/#CookieStoreManager).

We implement the subscribe, unsubscribe, and getSubscriptions
functions. The list of subscriptions is stored in the Service
Worker Registration on the Network Process side. The WebProcess
sends IPC messages to alter or get a copy of the list.

This patch causes some WPT Layout Tests that were previously
failing to now pass. Some of the tests whose results changed
to timeout are timing out because they are testing the
functionality of the ExtendableCookieChangeEvent for Service
Workers which has not yet been implemented. Since we currently
don't plan on firing cookie change events to the Service Workers,
there's no plan to implement ExtendableCookieChangeEvent yet.

* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStoreManager_getSubscriptions_empty.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStoreManager_getSubscriptions_empty.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStoreManager_getSubscriptions_multiple.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStoreManager_getSubscriptions_multiple.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStoreManager_getSubscriptions_single.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStoreManager_getSubscriptions_single.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_subscribe_arguments.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_subscribe_arguments.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/cookieStore_subscriptions_empty.https.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/serviceworker_cookieStore_subscriptions_reset.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/serviceworker_cookiechange_eventhandler_mismatched_subscription.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/serviceworker_cookiechange_eventhandler_multiple_subscriptions.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/serviceworker_cookiechange_eventhandler_overlapping_subscriptions.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/serviceworker_cookiechange_eventhandler_single_subscription.https.any.serviceworker-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/cookie-store/serviceworker_oncookiechange_eventhandler_single_subscription.https.any.serviceworker-expected.txt:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/cookie-store/CookieChangeSubscription.h: Added.
(WebCore::CookieChangeSubscription::CookieChangeSubscription):
(WebCore::CookieChangeSubscription::isHashTableDeletedValue const):
(WTF::CookieChangeSubscriptionHash::hash):
(WTF::CookieChangeSubscriptionHash::equal):
* Source/WebCore/Modules/cookie-store/CookieStoreGetOptions.idl:
* Source/WebCore/Modules/cookie-store/CookieStoreManager.cpp:
(WebCore::CookieStoreManager::create):
(WebCore::CookieStoreManager::CookieStoreManager):
(WebCore::CookieStoreManager::subscribe):
(WebCore::CookieStoreManager::unsubscribe):
(WebCore::CookieStoreManager::getSubscriptions):
* Source/WebCore/Modules/cookie-store/CookieStoreManager.h:
* Source/WebCore/Modules/cookie-store/CookieStoreManager.idl:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/workers/service/SWClientConnection.h:
* Source/WebCore/workers/service/ServiceWorkerContainer.cpp:
(WebCore::ServiceWorkerContainer::ensureProtectedSWClientConnection):
(WebCore::ServiceWorkerContainer::addCookieChangeSubscriptions):
(WebCore::ServiceWorkerContainer::removeCookieChangeSubscriptions):
(WebCore::ServiceWorkerContainer::cookieChangeSubscriptions):
* Source/WebCore/workers/service/ServiceWorkerContainer.h:
* Source/WebCore/workers/service/ServiceWorkerRegistration.cpp:
(WebCore::ServiceWorkerRegistration::cookies):
(WebCore::ServiceWorkerRegistration::addCookieChangeSubscriptions):
(WebCore::ServiceWorkerRegistration::removeCookieChangeSubscriptions):
(WebCore::ServiceWorkerRegistration::cookieChangeSubscriptions):
(WebCore::ServiceWorkerRegistration::protectedContainer const):
* Source/WebCore/workers/service/ServiceWorkerRegistration.h:
* Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
(WebCore::WorkerSWClientConnection::~WorkerSWClientConnection):
(WebCore::WorkerSWClientConnection::matchRegistration):
(WebCore::WorkerSWClientConnection::getRegistrations):
(WebCore::WorkerSWClientConnection::whenRegistrationReady):
(WebCore::WorkerSWClientConnection::scheduleUnregisterJobInServer):
(WebCore::WorkerSWClientConnection::subscribeToPushService):
(WebCore::WorkerSWClientConnection::unsubscribeFromPushService):
(WebCore::WorkerSWClientConnection::getPushSubscription):
(WebCore::WorkerSWClientConnection::getPushPermissionState):
(WebCore::WorkerSWClientConnection::getNotifications):
(WebCore::WorkerSWClientConnection::enableNavigationPreload):
(WebCore::WorkerSWClientConnection::disableNavigationPreload):
(WebCore::WorkerSWClientConnection::setNavigationPreloadHeaderValue):
(WebCore::WorkerSWClientConnection::getNavigationPreloadState):
(WebCore::WorkerSWClientConnection::startBackgroundFetch):
(WebCore::WorkerSWClientConnection::backgroundFetchInformation):
(WebCore::WorkerSWClientConnection::backgroundFetchIdentifiers):
(WebCore::WorkerSWClientConnection::abortBackgroundFetch):
(WebCore::WorkerSWClientConnection::matchBackgroundFetch):
(WebCore::WorkerSWClientConnection::retrieveRecordResponse):
(WebCore::WorkerSWClientConnection::retrieveRecordResponseBody):
(WebCore::WorkerSWClientConnection::addCookieChangeSubscriptions):
(WebCore::WorkerSWClientConnection::removeCookieChangeSubscriptions):
(WebCore::WorkerSWClientConnection::cookieChangeSubscriptions):
* Source/WebCore/workers/service/WorkerSWClientConnection.h:
* Source/WebCore/workers/service/server/SWServerRegistration.cpp:
(WebCore::SWServerRegistration::addCookieChangeSubscriptions):
(WebCore::SWServerRegistration::removeCookieChangeSubscriptions):
(WebCore::SWServerRegistration::cookieChangeSubscriptions const):
* Source/WebCore/workers/service/server/SWServerRegistration.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.cpp:
(WebKit::WebSWServerConnection::sharedPreferencesForWebProcess const):
(WebKit::WebSWServerConnection::addCookieChangeSubscriptions):
(WebKit::WebSWServerConnection::removeCookieChangeSubscriptions):
(WebKit::WebSWServerConnection::cookieChangeSubscriptions):
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.h:
* Source/WebKit/NetworkProcess/ServiceWorker/WebSWServerConnection.messages.in:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.cpp:
(WebKit::WebSWClientConnection::addCookieChangeSubscriptions):
(WebKit::WebSWClientConnection::removeCookieChangeSubscriptions):
(WebKit::WebSWClientConnection::cookieChangeSubscriptions):
* Source/WebKit/WebProcess/Storage/WebSWClientConnection.h:

Canonical link: https://commits.webkit.org/282417@main
